src/util/base.cc

```diff
@@ -0,0 +1,231 @@
+{
+  std::uint8_t in_buffer[Size + 1];
+  std::uint8_t out_buffer[Size * 2 + 1];
+  size_t ret(0);
+  memset(in_buffer, 0, sizeof(in_buffer));
+  memset(out_buffer, 0, sizeof(out_buffer));
+  do
+    {
+      input.read(reinterpret_cast<char*>(in_buffer), Size);
+      if (input.bad())
+        {
+          std::cerr << "Error while reading input ! " << strerror(errno)
+                    << std::endl;
+          return false;
+        }
+
+      if (encode)
+        ret = ByteStreamToBase(
+            in_buffer,
+            input.gcount(),
+            reinterpret_cast<char*>(out_buffer),
+            Size * 2);
+      else
+        ret = BaseToByteStream(
+            reinterpret_cast<char*>(in_buffer),
+            input.gcount(),
+            out_buffer,
+            Size * 2);
+
+      output.write(reinterpret_cast<char*>(out_buffer), ret);
+      if (output.bad())
+        {
+          std::cerr << "Error : Output to stream failed !" << std::endl;
+          return false;
+        }
+      output.flush();
+    } while (input.good());
+  return true;
+}
+
+bool BaseCommand::Impl(
+    const std::string& cmd_name,
+    int argc,
+    const char* argv[])
+{
+  bpo::variables_map vm;
+  bpo::positional_options_description pos;
+  pos.add("type", 1).add("infile", 1).add("outfile", 1);
+  try
+    {
+      bpo::parsed_options parsed = bpo::command_line_parser(argc, argv)
+                                       .options(m_desc)
+                                       .positional(pos)
+                                       .run();
+      bpo::store(parsed, vm);
+      bpo::notify(vm);
+    }
+  catch (const bpo::error& e)
+    {
+      std::cerr << GetName() << " ERROR: " << e.what() << std::endl;
+      return false;
+    }
+
+  if (vm.count("help") || (argc < 2) || (argc > 4))
+    {
+      if (argc < 2)
+        std::cerr << "Not enough arguments !" << std::endl;
+      else if (argc > 4)
+        std::cerr << "Too many arguments !" << std::endl;
+      PrintUsage(cmd_name);
+      return false;
+    }
+
+  // encode or decode or error
+  bool encode = true;
+  if (m_opt_type == "decode")
+    {
+      encode = false;
+    }
+  else if (m_opt_type != "encode")
+    {
+      std::cerr << "Invalid option : \"" << m_opt_type << "\"" << std::endl;
+      std::cerr << "Should be \"encode\" or \"decode\"" << std::endl;
+      PrintUsage(cmd_name);
+      return false;
+    }
+
+  // Streams management
+  std::shared_ptr<std::istream> istream;
+  if (m_opt_infile.empty() || m_opt_infile == "-")
+    {  // read from stdin
+      istream.reset(&std::cin, [](...) {});  // noop : don't delete
+    }
+  else
+    {  // read from file
+      std::ifstream* ifs = new std::ifstream();
+      ifs->open(m_opt_infile.c_str(), std::ios::in | std::ios::binary);
+      if (ifs->fail())
+        {
+          std::cerr << "Failed to open input file \"" << m_opt_infile.c_str()
+                    << "\"" << std::endl;
+          delete ifs;
```

Comment:
We're big on *smart pointers at all costs when not as a function parameter* (even though I see what you want, and it works with `shared_ptr`). Just an FYI for future design considerations.